src/core/AutomationPattern.cpp

```diff
@@ -417,33 +613,31 @@ float *AutomationPattern::valuesAfter( const TimePos & _time ) const
 
 
 
-void AutomationPattern::flipY( int min, int max )
+void AutomationPattern::flipY(int min, int max)
 {
-	timeMap tempMap = m_timeMap;
-	timeMap::ConstIterator iterate = m_timeMap.lowerBound(0);
-	float tempValue = 0;
+	QMutexLocker m(&m_patternMutex);
 
-	int numPoints = 0;
+	timeMap::iterator it = m_timeMap.lowerBound(0);
 
-	for( int i = 0; ( iterate + i + 1 ) != m_timeMap.end() && ( iterate + i ) != m_timeMap.end() ; i++)
-	{
-		numPoints++;
-	}
+	if (it == m_timeMap.end()) { return; }
 
-	for( int i = 0; i <= numPoints; i++ )
-	{
+	float tempValue = 0;
+	float outValueOffset = 0;
 
-		if ( min < 0 )
+	do
+	{
+		if (min < 0)
 		{
-			tempValue = valueAt( ( iterate + i ).key() ) * -1;
-			putValue( TimePos( (iterate + i).key() ) , tempValue, false);
+			it.value() *= -1.0;
 		}
 		else
 		{
-			tempValue = max - valueAt( ( iterate + i ).key() );
-			putValue( TimePos( (iterate + i).key() ) , tempValue, false);
+			tempValue = max - valueAt(POS(it));
+			outValueOffset = OFFSET(it) * -1.0;
+			putValues(TimePos(POS(it)), tempValue, tempValue + outValueOffset, false, true);
```

Comment:
Is it possible to replace this with an in-place operation if it doesn't lead to UB?

Reply:
I rewrote the `flipY` method as mentioned on Discord. The previous one expected that the range was either `[-max,+max]` or `[0,+max]`. If `min` was not `0` or `-max` the method could return us values that were out of range. I.e.: If `min = -5` and `max = 10`, if we had a node with value 8 it would become -8 falling out of range.

The new logic uses the distance between the node values and one of the limits (max) and apply that distance to the opposite limit (min). In that previous example, with `min = -5` and `max = 10`, if a node had value 8, the distance between the value and `max` is 2. Adding that distance to `min` gives us a new value of -3, which lies inside the range and is the proportional flipping of the value.

While rewriting it, I replaced the `while` loop with a `for` loop and used `m_timeMap.begin()` instead. I also added a boolean variable to check if any node was changed to avoid calling `generateTangents` and emitting a `dataChanged` signal unnecessarily.